### PR TITLE
Add Convex Hull in C

### LINKED ISSUE #3295: Add Convex Hull in C
To request a new code snippet, please fill out the following:

Project name: Convex Hull
Project link: https://sampleprograms.io/projects/convex-hull
Language: C

> The project link must be a link to a project on the [Sample Programs Website project list][sample-programs-project-list].
> If you would like to add a new project to the Sample Programs Website,
> please first make a pull request to the [Sample Programs Website][sample-programs-website].

[sample-programs-website]: https://github.com/TheRenegadeCoder/sample-programs-website
[sample-programs-project-list]: https://sampleprograms.io/projects/

### PR DESCRIPTION
**Pull request by 2Clutch**
## I Am Adding a New Code Snippet in an Existing Language

- [x] I fixed #3295
- [x] I named the pull request using `Add {PROJECT} in {LANGUAGE}` format

### CODE REVIEW

**Review by rzuckerm — CHANGES_REQUESTED** [review]
The error message needs to match the [project page](https://sampleprograms.io/projects/convex-hull/).

**Comment by rzuckerm** [comment]
See [this](https://github.com/TheRenegadeCoder/sample-programs/actions/runs/11564593566/job/32190077650?pr=3958#step:7:2943) for the test failures.

**Comment by 2Clutch** [comment]
> See [this](https://github.com/TheRenegadeCoder/sample-programs/actions/runs/11564593566/job/32190077650?pr=3958#step:7:2943) for the test failures.

I have a good idea of how to navigate this now so I should be fine (I hope). I have a solution in place for all of the issues I was assigned, and I'm now refactoring and making sure it follows the project description to the letter.

I don't want to start over so I suspect each PR will have 1 or 2 failures before going through.

**Review by rzuckerm — CHANGES_REQUESTED** [review]
The tests are still failing. See [this](https://github.com/TheRenegadeCoder/sample-programs/actions/runs/11565182047/job/32191764069?pr=3958#step:7:2943) for details. Also, please review the [project page](https://sampleprograms.io/projects/convex-hull/) again. The inputs are a list of X-coordinates and a list of Y-coordinates. The number of X- and Y-coordinates must be the same, and there must be at least 3 of them.

File: archive/c/c/convex-hull.c [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    Point *points = malloc(100 * sizeof(Point));  // Assume a maximum of 100 points
+    if (!points) {
+        printf("Memory allocation failed\n");
+        return 1;
+    }
+
+    int n = 0;
+    parseCoordinates(xStr, yStr, &points, &n);
+
+    printf("The points in the convex hull are:\n");
```
Comment (rzuckerm) [inline archive/c/c/convex-hull.c:116]: This `printf` needs to be removed

File: archive/c/c/convex-hull.c [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        printf("Different Cardinality\n");
+        exit(1);
+    }
+
+    *n = count;
+
+    if (count < 3) {
+        printf("Convex hull not possible with less than 3 points.\n");
+        exit(1);
+    }
```
Comment (rzuckerm) [inline archive/c/c/convex-hull.c:92]: The error message for both of these should be this:
```
Usage: please provide at least 3 x and y coordinates as separate lists (e.g. "100, 440, 210")
```

File: archive/c/c/convex-hull.c [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    if (val == 0) return 0;  // Collinear
+    return (val > 0) ? 1 : 2; // Clockwise or counterclockwise
+}
+
+// Function to compute the convex hull using the Gift Wrapping algorithm
+void convexHull(Point points[], int n) {
+    if (n < 3) {
+        printf("Convex hull not possible with less than 3 points.\n");
+        return;
+    }
```
Comment (rzuckerm) [inline archive/c/c/convex-hull.c:22]: Not sure this is actually needed since the number of coordinates was checked before calling this function.

**Review by rzuckerm — CHANGES_REQUESTED** [review]
Something about your parsing algorithm is incorrect since there are errors like `Invalid Integers` for valid inputs. Please see [this](https://github.com/TheRenegadeCoder/sample-programs/actions/runs/11581572902/job/32242651150#step:7:2983) for details

File: archive/c/c/convex-hull.c [hunk 4]
```diff
@@ -1,121 +1,134 @@
...
-    if (*s == '\0') return 0; // Empty after optional '-'
+    for (int i = 0; i < n; i++) {
+        printf("(%d, %d)\n", hull[i].x, hull[i].y);
+    }
+}
 
+bool isInteger(const char *s) {
     while (*s) {
-        if (!isdigit(*s)) return 0;
+        if (*s < '0' || *s > '9') {
```
Comment (rzuckerm) [inline archive/c/c/convex-hull.c:77]: Negative numbers need to be handled

File: archive/c/c/convex-hull.c [hunk 5]
```diff
@@ -1,121 +1,134 @@
...
-        xToken = strtok(NULL, ",");
-        yToken = strtok(NULL, ",");
+
+        tokenX = strtok(NULL, ",");
+        tokenY = strtok(NULL, ",");
     }
 
-    if (xToken || yToken) {
+    if (tokenX || tokenY) {
         printf("Different Cardinality\n");
```
Comment (rzuckerm) [inline archive/c/c/convex-hull.c]: The error message still needs to display the usage message.

File: archive/c/c/convex-hull.c [hunk 6]
```diff
@@ -1,121 +1,134 @@
...
     }
+
+    convexHull(points, count);
+    free(points);
 }
 
 int main(int argc, char *argv[]) {
     if (argc != 3) {
-        printf("Usage: please provide at least 3 x and y coordinates as separate lists (e.g. \"100, 440, 210\")\n");
+        printf("Usage: please provide two coordinate lists (x, y)\n");
```
Comment (rzuckerm) [inline archive/c/c/convex-hull.c]: The error message needs to be this for all cases:
```
Usage: please provide at least 3 x and y coordinates as separate lists (e.g. "100, 440, 210")
```

File: archive/c/c/convex-hull.c [hunk 7]
```diff
@@ -1,121 +1,134 @@
...
+    char *inputY = argv[2];
 
-    // Validate input for empty strings
-    validateInputs(xStr, yStr);
-
-    Point *points = malloc(100 * sizeof(Point));  // Assume a maximum of 100 points
-    if (!points) {
-        printf("Memory allocation failed\n");
+    if (strlen(inputY) == 0) {
+        printf("Missing Y\n");
```
Comment (rzuckerm) [inline archive/c/c/convex-hull.c:127]: Same comment here

**Review by rzuckerm — CHANGES_REQUESTED** [review]

**Review by rzuckerm — CHANGES_REQUESTED** [review]
The parsing still isn't right. The usage error message is being shown for all the cases. See [this](https://github.com/TheRenegadeCoder/sample-programs/actions/runs/11582362465/job/32245234628#step:7:2983)

**Review by rzuckerm — APPROVED** [review]
@2Clutch I used your parsing from Binary Search to fix this.

**Pull request merged**